blockchain/v2/processor.go

```diff
@@ -0,0 +1,185 @@
+package v2
+
+import (
+	"fmt"
+
+	"github.com/tendermint/tendermint/p2p"
+	tdState "github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/types"
+)
+
+type peerError struct {
+	priorityHigh
+	peerID p2p.ID
+}
+
+type pcDuplicateBlock struct {
+	priorityNormal
+}
+
+type bcBlockResponse struct {
+	priorityNormal
+	peerID p2p.ID
+	block  *types.Block
+	height int64
+}
+
+type pcBlockVerificationFailure struct {
+	priorityNormal
+}
+
+type pcBlockProcessed struct {
+	priorityNormal
+	height int64
+	peerID p2p.ID
+}
+
+type pcProcessBlock struct {
+	priorityNormal
+}
+
+type pcStop struct {
+	priorityNormal
+}
+
+type pcFinished struct {
+	priorityNormal
+	height       int64
+	blocksSynced int
+	state        tdState.State
+}
+
+func (p pcFinished) Error() string {
+	return "finished"
+}
+
+type queueItem struct {
+	block  *types.Block
+	peerID p2p.ID
+}
+
+type blockQueue map[int64]queueItem
+
+type pcState struct {
+	height       int64
+	queue        blockQueue
+	chainID      string
+	blocksSynced int
+	draining     bool
+	tdState      tdState.State
+	context      processorContext
+}
+
+func (ps *pcState) String() string {
+	return fmt.Sprintf("queue: %d draining: %v, \n", len(ps.queue), ps.draining)
+}
+
+// newPcState returns a pcState initialized with the last verified block enqueued
+func newPcState(initBlock *types.Block, tdState tdState.State, chainID string, context processorContext) *pcState {
+	var myID p2p.ID = "thispeersid" // XXX: this should be the nodes peerID
+	initItem := queueItem{block: initBlock, peerID: myID}
+
+	return &pcState{
+		height:       initBlock.Height,
+		queue:        blockQueue{initBlock.Height: initItem},
+		chainID:      chainID,
+		draining:     false,
+		blocksSynced: 0,
+		context:      context,
+		tdState:      tdState,
+	}
+}
+
+// nextTwo returns the next two unverified blocks
+func (state *pcState) nextTwo() (queueItem, queueItem, error) {
+	if first, ok := state.queue[state.height+1]; ok {
+		if second, ok := state.queue[state.height+2]; ok {
+			return first, second, nil
+		}
+	}
+	return queueItem{}, queueItem{}, fmt.Errorf("not found")
+}
+
+// synced returns true when only the last verified block remains in the queue
+func (state *pcState) synced() bool {
+	return len(state.queue) == 1
+}
```

Comment:
Not sure this is true. It seems that in the typical case we endup with two blocks in the queue:
- last synced block is N, queue size is 1
- we get bcBlockResponse for N+1 and N+2, queue size is 3
- we get and process pcProcessBlock event and process block N+1 and remove block N (according to advance() below), queue size is 2
- assume max height is N+2 and we are finished, we get pcStop and synced() returns false

It is also possible that last event in the fast sync is a peer error that causes the queue to be emptied and in this case queue length will be 0 (or 1 if start sync block N is still there as it has no peer)